File: 2021/3/solve.py
INPUT_FILE = '2021/3/input.txt'

def input(name = INPUT_FILE):
    with open(name) as f:
        for l in f:
            yield l.strip('\n')

def part1(i):
    freq = [0]*12
    t = 0

    for n in i:
        for g, b in enumerate(n):
            if b=='1':
                freq[g] += 1
        t += 1
    
    gamma = list(map(lambda g: '1' if g*2 >= t else '0', freq))
    epsilon = list(map(lambda g: '1' if g*2 <= t else '0', freq))

    print(int(''.join(gamma), 2)*int(''.join(epsilon), 2))

def oxco2(i, p):
    f = 0
    t = 0

    for n in i:
        # print(n)
        if n[p]=='1':
            f += 1
        t += 1
    
    oxigen = f*2 >= t
    co2 = f*2 < t

    return oxigen, co2

def part2(i):
    o = list(i)
    o_level = 0
    for p in range(12):
        fo = '1' if oxco2(o, p)[0] else '0'
        o = list(filter(lambda n: n[p]==fo, o))

        if len(o) == 1:
            print("O2")
            o_level = int(''.join(o[0]), 2)
            print(o_level)

    c = list(i)
    c_level = 0
    for p in range(12):
        fc = '1' if oxco2(c, p)[1] else '0'
        c = list(filter(lambda n: n[p]==fc, c))

        if len(c) == 1:
            print("CO2")
            c_level = int(''.join(c[0]), 2)
            print(c_level)

    print(o_level * c_level)

def solve():
    i = list(input())

    part1(i)
    part2(i)

solve()

File: 2021/3/solve.cpp
// requires -std=c++17

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <bitset>
using namespace std;

#define BITNUM 12

vector<bitset<BITNUM>> readinput()
{
    vector<bitset<BITNUM>> input;
    string filename = "input.txt";
    std::bitset<BITNUM> line;

    {
        ifstream file(filename, ios::in);
        while(file >> line)
            input.push_back(line);
    }

    return input;
}

void part1(vector<bitset<BITNUM>>& input)
{
    int t = input.size();
    vector<int> f = {0,0,0,0,0,0,0,0,0,0,0,0};
    std::bitset<BITNUM> gamma, epsilon;

    for(auto b: input)
        for(int i=0; i<BITNUM; i++)
            f[i] += b[i];

    for(int i=0; i<BITNUM; i++)
    {
        gamma[i] = f[i]*2 >= t;
        epsilon[i] = f[i]*2 <= t;
    }

    cout << gamma.to_ulong()*epsilon.to_ulong() << endl;
}

int oxco2(vector<bitset<BITNUM>>& input, vector<bool>& mask, int p)
{
    int f = 0, i = 0;
    for(auto b: input)
    {
        if(mask[i])
            f += b[p];
        i++;
    }

    return f;
}

int filter(vector<bitset<BITNUM>>& input, vector<bool>& mask, int p, bool fo)
{
    int t = 0, i = 0;
    for(auto b: input)
    {
        if(mask[i] && b[p]!=fo)
            mask[i] = false;
        if(!mask[i])
            t++;
        i++;
    }

    return input.size() - t;
}

int findlast(vector<bool>& mask)
{
    int i = 0;
    for(;!mask[i] && i<mask.size(); i++){}
    return i;
}

void part2(vector<bitset<BITNUM>>& input)
{
    int s = input.size();
    vector<bool> mask(s, true);
    int o_level;
    int c_level;

    int t = s;
    for(int p=BITNUM-1; t>1 && p>=0; p--)
    {
        int f = oxco2(input, mask, p);
        bool fo = f*2 >= t;

        t = filter(input, mask, p, fo);
    }
    o_level = input[findlast(mask)].to_ulong();

    mask = vector<bool>(s, true);
    t = s;
    for(int p=BITNUM-1; t>1 && p>=0; p--)
    {
        int f = oxco2(input, mask, p);
        bool fc = f*2 < t;

        t = filter(input, mask, p, fc);
    }
    c_level = input[findlast(mask)].to_ulong();

    cout << o_level*c_level << endl;
}

int main()
{
    vector<bitset<BITNUM>> input = readinput();
    part1(input);
    part2(input);

    return 0;
}
